### PR TITLE
Add sdpa with sinks

### PR DESCRIPTION
**Pull request by awni**
Not yet intended to merge yet (or maybe ever). But I want to make a PR for this so we can debate the trade-offs with something concrete.

See https://github.com/ml-explore/mlx-lm/pull/418 for corresponding change to gpt-oss in mlx-lm.

Prompt-processing time gets a nice speed bump:

For prompt length 2048 on M2U:

Pre: `prompt_tps=1299.417`
Post: `prompt_tps=1475.834`

### CODE REVIEW

File: mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h [hunk 1]
```diff
@@ -232,6 +234,15 @@ template <
     max_score[i] = Limits<AccumType>::finite_min;
   }
 
+  // TODO condition here is wrong
+  if (has_sinks) {
```
Comment (awni) [inline mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h]: This condition is not right, sinks need to be incorporated once per vector of qk scores.. I didn't figure it out yet.

Comment (angeloskath) [reply mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h]: Let me know if you want me to test this more but I think this should be

```cpp
if (has_sinks && sn == 0) {
}
```

basically once per row.

File: mlx/backend/metal/kernels/sdpa_vector.h [hunk 2]
```diff
@@ -268,6 +291,10 @@ template <typename T, int D, int V = D>
         score += q[i] * k[i];
       }
       score = simd_sum(score);
+      if (score < Limits<T>::finite_min) {
```
Comment (angeloskath) [inline mlx/backend/metal/kernels/sdpa_vector.h:292]: Interesting does that fix a bug somewhere?

**Review by angeloskath — COMMENTED** [review]
This looks great. I think I left a comment that fixes the batch case if not let me know and I will investigate further.

File: mlx/backend/metal/kernels/sdpa_vector.h (hunk 2 above)
Comment (awni) [reply mlx/backend/metal/kernels/sdpa_vector.h:292]: Yea it's very much an edge case but if `q_i^T k_j = -inf` for any pair then previously it would create NaNs, now it will do the right thing. I added a test for it as well.

File: mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h (hunk 1 above)
Comment (awni) [reply mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h]: It sounds like it should be right.. but it doesn't pass the test case I added for this 🤔

Comment (awni) [reply mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h]: As it turns out the original condition is correct (it needs to be applied for every thread).. and the problem was the log2(e) factor.

**Comment by awni** [comment]
Ok I think this is ready to land once it's approved. Lmk if there are any other comments.

File: mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h (hunk 1 above)
Comment (angeloskath) [reply mlx/backend/metal/kernels/steel/attn/kernels/steel_attention.h]: Sorry for the noise then :-)
